### PR TITLE
fix(epic): structured OperationOutcome detection for Epic 400 scope-rejection

### LINKED ISSUE #1096: epic-connector: structured detection of Epic 400 scope-rejection (replace string-match)
## Context
Discovered during review of PR #1095.

PR #1095 introduces `isUnauthorizedSubResourceError(err)` in `services/epic-connector/src/sync/sync-jobs.ts`, which matches on substrings of `Error.message`:

```ts
return (
  err.message.includes("Combination of parameters is not valid") ||
  err.message.includes("authorized sub-resource")
);
```

This works against Epic's R4 sandbox today, but is fragile:

1. **Vendor wording is unstable.** Epic can (and historically does) reword `OperationOutcome.diagnostics` between sandbox refreshes — any wording change silently turns soft-skips into hard failures.
2. **Other unrelated 400s can collide.** Any future 400 whose body happens to contain the phrase "Combination of parameters" (e.g. a developer mis-spell on a non-Epic upstream) would be silently swallowed.
3. **OperationOutcome carries a real code.** Epic emits `OperationOutcome.issue[].code = "invalid"` and the diagnostics text references error code `59022`. Today the fhir-client flattens that to an `Error` and we re-parse the string.

## Proposal
Introduce a structured Epic error class (or discriminated `EpicError` union) thrown by `fhir-client.ts`, carrying:
- `status` (HTTP code)
- `operationOutcomeCode` (e.g. `"invalid"`, `"forbidden"`, `"security"`)
- `operationOutcomeId` (e.g. Epic's `59022` numeric error code, when present)
- `diagnostics` (the original text)

Then `isUnauthorizedSubResourceError` becomes:
```ts
err instanceof EpicError && (err.operationOutcomeId === 59022 /* or similar */)
```

## Acceptance
- [ ] `fhir-client.ts` throws a typed `EpicError` carrying parsed OperationOutcome metadata
- [ ] `isUnauthorizedSubResourceError` switches to structural detection
- [ ] String-match fallback retained one release as a belt-and-braces, with a `log.warn` if the structural path didn't match but the string did
- [ ] Unit test covering: structural detection matches; non-Epic 400 with similar wording does NOT match; Epic 5xx does NOT match

## Refs
- PR #1095
- `services/epic-connector/src/sync/sync-jobs.ts` (current detector)
- `services/epic-connector/src/fhir-client.ts` (operationOutcomeError)

### LINKED ISSUE #1099: epic-connector: add tests for OperationOutcome-format error detection + non-active MedicationRequest history
## Context
Discovered during review of PR #1095.

PR #1095 adds 7 tests, which is solid coverage. Two gaps worth filling:

1. **OperationOutcome-format error**. The Epic fhir-client throws two different error-message shapes (see `fhir-client.ts`):
   - `Epic FHIR request returned 400 Bad Request — <body>` (from `requestWithRetry`)
   - `Epic returned OperationOutcome: invalid — <diagnostics>` (from `operationOutcomeError()`)

   Today's tests only cover the first shape. `isUnauthorizedSubResourceError` matches on the diagnostics substring so it _should_ work for both, but there's no test that locks that in. A future refactor to either error-construction path could silently break soft-skip detection.

2. **Non-active MedicationRequest history**. The hardcoded `status=active` filter means stopped/cancelled meds are never imported. The PR explicitly defers this to a future iteration. A test (currently `xit`-skipped) documenting the expected future shape (override mechanism, multi-status fan-out) would lock in the contract.

3. **Multiple categories returning duplicates**. The dedup test covers one shared Observation across two categories — but only when both categories return ONLY the shared resource. A test covering "vital-signs returns [A, B]; laboratory returns [B, C]; final result is [A, B, C]" would strengthen the dedup invariant.

4. **`searchAll` thrown _synchronously_ vs from inside the async generator**. Test "re-throws errors that are NOT auth-scope failures" throws synchronously from `searchAll`. The "swallows" test throws from inside the async generator. Both paths should exist for the soft-skip case too (synchronous Epic-format auth error vs in-generator).

## Acceptance
- [ ] Test: `isUnauthorizedSubResourceError` matches an `Error` whose message starts with `"Epic returned OperationOutcome: invalid —"` and contains the soft-fail diagnostics
- [ ] Test: Observation dedup over `[A,B] + [B,C] = [A,B,C]`
- [ ] Test: soft-skip works when the auth-scope Error is thrown synchronously from `searchAll` (not just from the async generator body)
- [ ] (Documentation-only) Add an `xit` placeholder for future multi-status MedicationRequest fan-out

## Refs
- PR #1095
- `services/epic-connector/src/__tests__/sync-jobs.test.ts`

### PR DESCRIPTION
**Pull request by blamechris**
## Summary
- Introduces `EpicFhirError extends Error` carrying the parsed `OperationOutcome` so consumers can discriminate Epic failure categories structurally (preferred) instead of substring-matching the human text (#1095's interim fix).
- Rewrites the sync-jobs auth-scope detector to prefer `err.hasIssueCode("59022")` on the typed error, with the substring match retained as a safety fallback for cases where Epic returns a non-OperationOutcome body.
- Adds 7 new sync-jobs tests (+ 1 `it.skip` placeholder) and 3 new fhir-client tests covering the structured + fallback + edge-case paths.

Closes #1096. Closes #1099.

## Why this matters
PR #1095 landed a working soft-skip but matched on substrings — fragile to Epic rewording. This PR closes that loop with proper FHIR-spec discriminators:

```ts
// before (#1095): substring on Error.message
if (err.message.includes("authorized sub-resource")) return true;

// after (#1096): structured check on parsed OperationOutcome
if (err instanceof EpicFhirError && err.hasIssueCode("59022")) return true;
// substring kept as fallback for non-JSON Epic bodies (proxy 502, etc.)
if (err instanceof Error && err.message.includes("authorized sub-resource")) return true;
```

The Epic error-code table the connector now reasons about:

| Code | Meaning | Action |
|---|---|---|
| `59022` | "Combination of parameters is not valid for any authorized sub-resource" | Soft-skip (scope-rejection) |
| `59108` | "A required element is missing" | Surface (request-shape bug) |

## Coverage added per #1099 acceptance criteria
- [x] Structured detection: `EpicFhirError` with code 59022 → soft-skipped
- [x] Structured detection: `EpicFhirError` with code 59108 → surfaced (NOT swallowed)
- [x] Substring fallback when `EpicFhirError.operationOutcome` is undefined but message contains "authorized sub-resource"
- [x] Substring match on the alternate `"Epic returned OperationOutcome:"` shape emitted by `operationOutcomeError()` (acceptance #1)
- [x] Observation dedup invariant over `[A,B] + [B,C] → [A,B,C]` (acceptance #2)
- [x] Soft-skip works when `EpicFhirError` is thrown **synchronously** from `searchAll`, not just from inside the async-generator body (acceptance #3)
- [x] `it.skip` placeholder documenting future multi-status MedicationRequest fan-out (acceptance #4)

## fhir-client changes
- `EpicFhirError` exposes `status`, `statusText`, `body` (raw), `operationOutcome` (parsed or undefined), and `hasIssueCode(code)`.
- `requestWithRetry` now throws `EpicFhirError` from both the 5xx-after-max-retries path and the 4xx-non-401-non-429 path.
- `OperationOutcomeIssue` and `OperationOutcomeIssueCoding` types added; `operationOutcomeHasIssueCode(outcome, code)` helper exported for non-error callsites.

## Test plan
- [x] `pnpm --filter @carebridge/epic-connector test` — 145 pass + 1 documented skip (16 test files, including 7 new sync-jobs tests + 3 new fhir-client tests)
- [x] `pnpm typecheck` — full workspace, 38/38 clean (no breakage in downstream consumers of fhir-client errors)
- [x] Live verify against `fhir.epic.com` sandbox (Camila Lopez `erXuFYUfucBZaryVksYEcMg3`) — 5/5 resource types green via the structured detection path; vital-signs + MedicationRequest soft-skipped via `hasIssueCode("59022")` (not substring)

### CODE REVIEW

**Comment by blamechris** [comment]
## Code Review: PR #1100

### Summary
Clean follow-on to #1095 that closes the substring-fragility loop by introducing `EpicFhirError` carrying a parsed `OperationOutcome`, and rewriting `isUnauthorizedSubResourceError` to discriminate Epic 400s structurally via `details.coding[].code === "59022"`. The substring fallback is retained as a defensive safety net for non-FHIR error bodies (HTML 502 from fronting proxies), which is the right call — Epic edge layers do occasionally bypass FHIR formatting. Tests cover the full matrix per #1099's acceptance criteria: structured detection (59022 swallowed, 59108 surfaced), substring fallback for both `EpicFhirError`-with-undefined-OO and the alternate `"Epic returned OperationOutcome:"` shape, the dedup invariant `[A,B] + [B,C] -> [A,B,C]`, and the synchronous-throw path through `collectSearchOrSkipUnauthorized`.

### Strengths
- **Structural beats substring.** The new path inspects `OperationOutcome.issue[].details.coding[].code` which is the actual FHIR-spec discriminator. Epic re-wording the human text won't break this.
- **Defensive substring fallback retained with justification.** The comment explicitly calls out the "edge proxy returns HTML/plain text" case and why the more-specific `"authorized sub-resource"` phrase is used (not the looser `"Combination of parameters is not valid"` prefix that would also fire for 59108).
- **Negative test for 59108.** The "MUST NOT swallow" test for the missing-required-element code is the most important test in the suite — it locks in the policy that future code can't accidentally start swallowing real bugs.
- **Synchronous-throw test (#1099 acceptance #3).** Catches a subtle bug class where the try/catch around `searchAll` only wraps the iterator-body throws, not the `searchAll()` call itself.
- **Dedup invariant test (#1099 acceptance #2).** Asserts on persisted IDs (`["A","B","C"]`), not just count — much stronger than a count assertion alone.
- **`body` field on `EpicFhirError`** keeps the raw response for diagnostic logging when the OperationOutcome parse fails. `safeReadBody` already caps at 500 chars (`fhir-client.ts:475`), so memory pressure isn't a concern.
- **Backward-compat preserved.** `EpicFhirError extends Error` and the `message` format is unchanged from the pre-PR string. Existing log consumers reading `err.message` continue to work.
- **Public API surface is well-curated.** `index.ts` re-exports `EpicFhirError`, `OperationOutcomeIssue`, `OperationOutcomeIssueCoding`, and the helper `operationOutcomeHasIssueCode` — downstream services can import the same primitives without reaching into private paths.

### Issues Found

#### Critical
_None._ The diff is correct, type-safe, and the test coverage is solid.

#### Suggestions

| # | Issue | Why |
|---|---|---|
| S1 | `operationOutcomeError()` still throws plain `Error`, not `EpicFhirError` | Used by `read()` / `createResource()` / `updateResource()` (lines 157, 245, 278) when Epic returns 200 + an OperationOutcome body. The structured `hasIssueCode("59022")` check falls through to the substring fallback for these paths, partially undermining the PR's goal. Tracked as #1101. |
| S2 | `EPIC_UNAUTHORIZED_SUB_RESOURCE_CODE` as a lone const | `EpicFhirError` docstring already references 59108 as the next code the connector will need. Extracting an `EPIC_ERROR_CODES` registry now (with symbolic names) avoids string-literal duplication and makes future code-based checks discoverable via autocomplete. Tracked as #1102. |
| S3 | `tryParseOperationOutcome` lacks direct unit tests | The function is the trust boundary for promoting raw bytes to typed OperationOutcome. Covered by integration tests for happy + HTML-body cases, but no direct tests for `"null"`, `[1,2,3]`, valid JSON without `resourceType`, valid JSON with wrong `resourceType`, or OperationOutcome with `issue` missing/non-array. `JSON.parse` itself is safe (no proto pollution risk), but the `as OperationOutcome` assertion is unchecked beyond `resourceType`. Tracked as #1103. |
| S4 | `OperationOutcomeIssueCoding.code?: string` is optional | FHIR R4 permits it, but Epic always emits it in practice. A coding without `code` carries no useful info for structured discrimination — `hasIssueCode` silently returns false rather than flagging the data quality issue. Tracked as #1104. |

#### Nitpicks

| # | Issue | Why |
|---|---|---|
| N1 | `it.skip(...)` placeholder with `expect(false).toBe(true)` body | Satisfies #1099 acceptance #4, but `it.skip` tends to rot and the bare `expect(false)` is a footgun if accidentally un-skipped. Consider `expect.fail("see #NNNN")` or a top-of-file TODO comment instead. Tracked as #1105. |
| N2 | `class EpicFhirError extends Error` vs functional style | CLAUDE.md says "no classes unless necessary." Extending `Error` IS the canonical TypeScript pattern (a tagged interface won't give you `err instanceof EpicFhirError` discrimination, stack traces, or proper async error propagation through the runtime). This is the right exception. No action needed — flagging only because the PR review brief asked. |
| N3 | `hasIssueCode` does O(issues x codings) scan | Realistic OperationOutcomes have 1-3 issues with 1-3 codings each. Not a hot path. No action. |

### Deferred Items

| # | Description | Issue |
|---|---|---|
| D1 | Unify `operationOutcomeError()` to throw `EpicFhirError` | https://github.com/blamechris/carebridge/issues/1101 |
| D2 | Extract `EPIC_ERROR_CODES` registry | https://github.com/blamechris/carebridge/issues/1102 |
| D3 | Direct unit tests for `tryParseOperationOutcome` edge cases | https://github.com/blamechris/carebridge/issues/1103 |
| D4 | Tighten `OperationOutcomeIssueCoding.code` typing | https://github.com/blamechris/carebridge/issues/1104 |
| D5 | Replace `it.skip` placeholder with TODO / `expect.fail` | https://github.com/blamechris/carebridge/issues/1105 |

### Architecture Notes

- **Error-type discrimination is the right primitive.** Inter-service boundary errors should always be typed, not stringly-discriminated. This PR sets the pattern for the rest of the FHIR connector and for future external-vendor integrations (Cerner, Allscripts) that will face the same vendor-wording fragility.
- **Two-tier detection is correct.** Structured-first with substring fallback acknowledges that real-world edge layers don't always preserve FHIR formatting. The fallback is appropriately narrow (`"authorized sub-resource"` phrase, not the broader prefix) so it can't silently swallow 59108 missing-element bugs.
- **The remaining inconsistency is `operationOutcomeError()`.** That function pre-dates this PR and lives in a different code path (2xx HTTP + OperationOutcome body), but for the structural-detection contract to be uniform across all Epic failure surfaces, it needs the same treatment. Captured in #1101.
- **No HIPAA/PHI concerns.** `EpicFhirError.body` keeps response text (500-char cap via `safeReadBody`), which is OperationOutcome diagnostics — no patient identifiers in the failure path.
- **No event-emission concerns.** This is the read/sync path; the BullMQ clinical-events queue invariant is upstream of this code.

### Verdict

**Approve.** No critical issues. The five deferred items are all incremental hardening on top of an already-correct PR — none block merge. Recommend merging as-is and picking up #1101 next since it closes the last remaining inconsistency in error typing.

File: services/epic-connector/src/fhir-types.ts [hunk 1]
```diff
@@ -49,15 +49,49 @@ export interface FhirResource {
...
+export function operationOutcomeHasIssueCode(
+  outcome: OperationOutcome,
+  code: string,
+): boolean {
+  for (const issue of outcome.issue) {
+    for (const c of issue.details?.coding ?? []) {
+      if (c.code === code) return true;
+    }
+  }
+  return false;
```
[the remark this answers is not in the document]
Comment (blamechris) [reply services/epic-connector/src/fhir-types.ts:100]: **FIX**

Fixed in `8646d4b`

**Change:** Added `Array.isArray(outcome.issue)` guard so `hasIssueCode` returns `false` (not throws) on malformed/partial OperationOutcome payloads. Regression test in `fhir-client.test.ts` drives the malformed-shape path (4xx with `{resourceType: "OperationOutcome", issue: null}`).

```diff
  export function operationOutcomeHasIssueCode(
    outcome: OperationOutcome,
    code: string,
  ): boolean {
+   if (!Array.isArray(outcome.issue)) return false;
    for (const issue of outcome.issue) {
      for (const c of issue.details?.coding ?? []) {
        if (c.code === code) return true;
      }
    }
    return false;
  }
```

Stricter validation inside `tryParseOperationOutcome` (your alternate suggestion) is tracked separately as #1103 — the array guard above is the immediate hardening; the structural validator is the longer-term cleanup.

File: services/epic-connector/src/fhir-client.ts [hunk 2]
```diff
@@ -24,9 +24,72 @@ import type {
...
+ *            sub-resource. Used by sync-jobs.ts to soft-skip categories
+ *            the registered app doesn't carry a scope for.
+ *
+ *   "59108"  A required element is missing (e.g. Observation without
+ *            category or code). Indicates a request-shape bug, NOT
+ *            a scope issue — must surface, not be swallowed.
+ *
+ * `body` keeps the raw response text for diagnostic logging when the
+ * response wasn't parseable as an OperationOutcome (e.g. Epic returned
+ * an HTML 502 page from a fronting proxy).
```
[the remark this answers is not in the document]
Comment (blamechris) [reply services/epic-connector/src/fhir-client.ts]: **FIX**

Fixed in `8646d4b`

**Change:** Updated the class doc + added an inline property doc so callers don't assume `body` is the full payload.

```diff
- * `body` keeps the raw response text for diagnostic logging when the
- * response wasn't parseable as an OperationOutcome (e.g. Epic returned
- * an HTML 502 page from a fronting proxy).
+ * `body` keeps a truncated snippet (first 500 chars via {@link safeReadBody})
+ * of the raw response text for diagnostic logging when the response wasn't
+ * parseable as an OperationOutcome (e.g. Epic returned an HTML 502 page
+ * from a fronting proxy). NOT the full payload — callers needing the
+ * complete body must re-fetch.
  */
  export class EpicFhirError extends Error {
    ...
+   /** Truncated body snippet (≤500 chars). Do NOT assume full payload. */
    public readonly body: string,
```

**Comment by blamechris** [comment]
## Review Comments Addressed

| # | Comment | Outcome | Reference |
|---|---------|---------|-----------|
| 1 | `operationOutcomeHasIssueCode` could throw on malformed `outcome.issue` | FIX | `8646d4b` |
| 2 | `body` doc comment claims raw text but `safeReadBody` truncates to 500 chars | FIX | `8646d4b` |

**Total:** 2 comments addressed
- Fixed: 2 (commit `8646d4b`)
- False positives: 0
- Follow-up issues created: 0 (related structural-validator concern in #1099 already tracked as #1103 from agent-review)
- Existing issues closed: 0

Tests: 146 pass + 1 documented skip (1 new regression test added for malformed-OperationOutcome path).
